File: server/src/main/java/com/fluffy/global/web/cookie/CookieManager.java
package com.fluffy.global.web.cookie;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import java.util.Arrays;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseCookie;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class CookieManager {

    private final CookieProperties cookieProperties;

    public ResponseCookie createAccessTokenCookie(String accessToken) {
        return ResponseCookie.from(cookieProperties.accessTokenKey(), accessToken)
                .httpOnly(cookieProperties.httpOnly())
                .secure(cookieProperties.secure())
                .domain(cookieProperties.domain())
                .path(cookieProperties.path())
                .maxAge(cookieProperties.maxAge())
                .build();
    }

    public ResponseCookie createExpiredAccessTokenCookie() {
        return ResponseCookie.from(cookieProperties.accessTokenKey(), "")
                .httpOnly(cookieProperties.httpOnly())
                .secure(cookieProperties.secure())
                .domain(cookieProperties.domain())
                .path(cookieProperties.path())
                .maxAge(0)
                .build();
    }

    public String extractAccessToken(HttpServletRequest request) {
        String tokenCookieName = cookieProperties.accessTokenKey();
        Cookie[] cookies = request.getCookies();

        if (cookies == null) {
            return null;
        }

        return Arrays.stream(cookies)
                .filter(cookie -> tokenCookieName.equals(cookie.getName()))
                .map(Cookie::getValue)
                .findFirst()
                .orElse(null);
    }
}

File: server/src/main/kotlin/com/fluffy/global/web/cookie/CookieManager.kt
package com.fluffy.global.web.cookie

import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.ResponseCookie
import org.springframework.stereotype.Component

@Component
class CookieManager(
    private val cookieProperties: CookieProperties,
) {

    fun createAccessTokenCookie(accessToken: String): ResponseCookie {
        return ResponseCookie.from(cookieProperties.accessTokenKey, accessToken)
            .httpOnly(cookieProperties.httpOnly)
            .secure(cookieProperties.secure)
            .domain(cookieProperties.domain)
            .path(cookieProperties.path)
            .maxAge(cookieProperties.maxAge)
            .build()
    }

    fun createExpiredAccessTokenCookie(): ResponseCookie {
        return ResponseCookie.from(cookieProperties.accessTokenKey, "")
            .httpOnly(cookieProperties.httpOnly)
            .secure(cookieProperties.secure)
            .domain(cookieProperties.domain)
            .path(cookieProperties.path)
            .maxAge(0)
            .build()
    }

    fun extractAccessToken(request: HttpServletRequest): String? {
        val tokenCookieName = cookieProperties.accessTokenKey
        val cookies = request.cookies

        if (cookies.isNullOrEmpty()) {
            return null
        }

        return cookies.firstOrNull { it.name == tokenCookieName }?.value

    }
}